models/particles/d3q27_PSM/Dynamics.c.Rt

```diff
@@ -0,0 +1,1067 @@
+
+        for (ParticleIterator p(X,Y,Z); p; ++p) if ((p.dist - p.rad)<1.5) {
+                if ((p.dist - p.rad) < - 1){
+                        coverage = 1.0;
+                } else{
+                        coverage = (p.rad -  0.084/p.rad + 0.5 - p.dist);
+                }
+
+                if (coverage > 1.0){ coverage =  1.0;}
+                if (coverage < 0.0){ coverage =  0.0;}
+
+                ret += coverage;
+
+        }
+
+        if (ret > 1.0){ ret =  1.0;}
+        if (ret < 0.0){ ret = 0.0;}
+
+        return ret;
+}
+
+CudaDeviceFunction real_t CalcPeriodicSolid(){
+        real_t coverage = 0, ret = 0;
+	
+        for (FullParticleIterator p(X,Y,Z); p; ++p) {
+	    real_t dx = p.diff.x;
+	    real_t dy = p.diff.y;
+	    real_t dz = p.diff.z;
+            if ((DNx>0) && ((DNx - abs(dx)) < (p.rad + 2))) {
+	        dx = DNx - abs(dx);
+            }
+	    
+	    if ((DNy>0) && ((DNy - abs(dy)) < (p.rad + 2))) {
+	        dy = DNy - abs(dy);
+            }
+	    		
+	    if ((DNz>0) && ((DNz - abs(dz)) < (p.rad + 2))) {
+	        dz = DNz - abs(dz);
+            }
+
+	    real_t dist = sqrt(dx*dx + dy*dy + dz*dz);
+	
+	    if ((dist - p.rad)<sqrt(3)) {
+                if ((dist - p.rad) < - 1){
+                        coverage = 1.0;
+                } else{
+                        coverage = (p.rad -  0.084/p.rad + 0.5 - dist);
+                }
+
+                if (coverage > 1.0){ coverage =  1.0;}
+                if (coverage < 0.0){ coverage =  0.0;}
+
+                ret += coverage;
+
+           }
+	}
+	
+	if (ret > 1.0){ ret =  1.0;}
+        if (ret < 0.0){ ret = 0.0;}
+
+        return ret;
+}
+
+CudaDeviceFunction void CalcF() {
+	 real_t numP = 0;
+	
+	    real_t omegaS[27], feqP[27], feq[27], u[3], uP[3];
+	    real_t d, usq, uPsq;
+	    
+	    uP[0] = 0.0;
+	    uP[1] = 0.0;
+	    uP[2] = 0.0;
+
+	    d=getRho();
+	    u[0] = ((<?R C(f %*% U[,1]) ?> )/d + 0.5*(AccelX + GravitationX) );
+	    u[1] = ((<?R C(f %*% U[,2]) ?> )/d + 0.5*(AccelY + GravitationY) );
+	    u[2] = ((<?R C(f %*% U[,3]) ?> )/d + 0.5*(AccelZ + GravitationZ) );
+
+                <?R
+                         u = PV(paste("u[",1:3-1,"]",sep=""))
+                         uP = PV(paste("uP[",1:3-1,"]",sep=""))
+                         d = PV("d")
+                         usq = PV ("usq")
+                         uPsq = PV ("uPsq")
+                         feq = PV(paste("feq[",1:27-1,"]",sep=""))
+                         feqP = PV(paste("feqP[",1:27-1,"]",sep=""))
+                         omegaS = PV(paste("omegaS[",1:27-1,"]",sep=""))
+                         omegaF = PV("omegaF")
+
+                      C(usq, sum(u*u)*(1/c_squ))
+                      for (i in 1:27){
+                         C( feq[i],  (((u %*% t(U) * (1/c_squ))[i]) + (((u %*% t(U) * (1/c_squ))[i])*((u %*% t(U) * (1/c_squ))[i])+usq*(-1))*0.5 + 1) * d *tn[i,i])
+                      }
+	 ?>
+
+         for (FullParticleIterator p(X,Y,Z); p; ++p) {
```

Comment:
The full particle iterator is inefficient. Please change this to `ParticleIterator` and check if your results are the same.